programs/survival/villager_auto_trader.sc

```diff
@@ -1,27 +1,57 @@
 
 global_trade_speed=120;//Trade every 120 ticks, so u have time to retrieve item. Same as piglin bartering, so seemed fair
 
-__on_tick()->(
-    if(tick_time()%global_trade_speed==2,
-        for(filter(entity_selector('@e[type=zombie_villager]'),nbt(_~'nbt':'HandItems'):'[]':0:'id'=='minecraft:emerald'),
-            hand_count=nbt(_~'nbt':'HandItems'):'[]':0:'Count';
-            if((trade_list = nbt(_~'nbt':'Offers':'Recipes'):'[]')==0,continue());
-            if(type(trade_list)!='list',trade_list=[trade_list]);
-            trade=first(trade_list,
-                _:'buy':'id'=='minecraft:emerald' &&//only emerald trades
-                _:'buyB':'id'=='minecraft:air' &&//don't want second trade item
-                _:'maxUses'!=0 &&//Cos if not it's cheaty, dont work rn
-                (cost=_:'buy':'Count')<=hand_count//Need to have enough emeralds
-            );
-            if(!trade,continue());
-
-            modify(_,'nbt_merge',str('{HandItems:[{id:"minecraft:emerald",Count:%sb}]}',hand_count-cost));
-
-            spawn('item',pos(_),str('{Item:{id:"%s",Count:%db,PickupDelay:0}}',trade:'sell':'id',trade:'sell':'Count'));
+entity_load_handler('zombie_villager', _(e, new) -> _trade(zombie_villager, null));
+
+
+_trade(zombie_villager, trade)->(
+    if(trade==null,//call this again in case the player messes with the zombie villager's nbt, also cos too lazy to shift elsewhere
+        if((trade_list = nbt(zombie_villager~'nbt':'Offers':'Recipes'):'[]')==0, return());//this return call would only be called once, so it's not expensive
+        trade=first(trade_list,
+            _:'buy':'id'=='minecraft:emerald' &&//only emerald trades
+            _:'buyB':'id'=='minecraft:air' &&//don't want second trade item
+            _:'maxUses'!=0 //Cos if not it's cheaty, dont work rn
+        );
+        if(!trade, return())//again, this is a cheap call and only called once
+    );
+
+    if(nbt(zombie_villager~'nbt':'HandItems'):'[]':0:'id'=='minecraft:emerald' && (hand_count=nbt(zombie_villager~'nbt':'HandItems'):'[]':0:'Count') >= ((cost=trade:'buy':'Count')),
+
+        modify(zombie_villager,'nbt_merge',str('{HandItems:[{id:"minecraft:emerald",Count:%sb}]}',hand_count-cost));
+
+        item = null;
+        in_dimension(zombie_villager,
+            item_pos = pos(zombie_villager)+[0, zombie_villager~'eye_height'-0.3, 0]; 
+            item = spawn('item',item_pos,str('{Item:{id:"%s",Count:%db, PickupDelay:10}}',trade:'sell':'id',trade:'sell':'Count'));//todo add velocity towards player
+        );
+
+        nearest_player = null;
+        distance_sq_to_nearest_player=25;
+
+        for(entity_area('player', item_pos, [5,5,5]),
+            distance_sq = _euclidean_sq(item_pos, pos(_));
+            if(distance_sq<distance_sq_to_nearest_player,
+                nearest_player = _;
+                distance_sq_to_nearest_player = distance_sq
+            )
+        );
+
+        if(nearest_player!=null,
```

Comment:
If no payer is around, i think the item should still be thrown in the zombie's look direction.